### ISSUE TITLE
Configure kubelet directory path differently for each worker nodes pool

### ISSUE THREAD

**Issue by ccojocar** [kubernetes-sigs/security-profiles-operator#1446]
#### What would you like to be added:

Add an option which allows to configure the kubelet directory differently for each worker nodes pool. 

A possible solution would be to request that all worker nodes with a different kubelet location are labeled for instance with `kubelet.kubernetes.io/directory-location` that has the kubelet directory path in the value. As an example for a node which is configured to use the kubelet directory path `/mnt/resource/kubelet`, the node should be labeled with:
```
kubelet.kubernetes.io/directory-location: mnt_resource_kubelet
```

The operator will list all nodes and identify all the distinct location labels. For each label, the operator will create a different daemonset object with a custom selector that match the value of the location label. It will also update the KUBELET_DIR environment variable of the daemonset to point to the same kubelet directory as the label value. 

The custom selector will ensure that the pods of each daemonset will be correctly scheduled on nodes that have the same kubelet directory as configured in their `KUBELET_DIR` environment variable.

#### Why is this needed:

The kubelet directory can be configured differently for each worker node and more typically for each worker nodes pool. The spo at the moment only allows to configure a custom kubelet directory for entire cluster. There are situations when the kubelet directory is different across the nodes in a cluster.

**Comment by saschagrunert** [comment]
This sounds like a custom downward API or such. I'm not sure about the label name itself but I think the feature is definitely worth to elaborate.

**Comment by ccojocar** [comment]
The[ downward API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/) seems like an interesting idea.The only node related information available is `nodeName` and `hostIP`. 

The challenge is that a daemonset pod will need to figure out dynamically the configured kubelet root directory of the node is running on.

The kubelet root directory is configured in[ `--root-dir`](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/#options) command line option. This values does't seem to be exposed anywhere. This is why requiring an additional label with this value in the node object somehow makes sense. The cluster operators can add such a value when they create a cluster which has nodes with a different kubelet root directory. 

In this case, an alternative to using a custom selector as proposed above, it would be to propagate down the `nodeName` to daemon pod containers. The `non-root-enabler` container can retrieve the node object and read kubelet root directory form the label. This will require an api-server access from non-root-enabler container.  

@saschagrunert How does sound to you?

**Comment by saschagrunert** [comment]
> This will require an api-server access from non-root-enabler container.

Hm yeah that sounds reasonable.  :+1:

**Comment by JAORMX** [comment]
I think we're starting to head in the direction of needing to support multiple SPOd deployments. Having to add a label to nodes and custom logic to translate the location names into something the SPO can understand seems error-prone to me. Instead, let's acknowledge that there is the possibility of heterogeneous nodes in clusters and we should cater for such use-cases.

We could instead have the kubelet folder location be a paramter in the SPOD object, and add the ability to schedule the daemonset in a subset of nodes via a custom selector.

Here's an example:

```yaml
---
apiVersion: security-profiles-operator.x-k8s.io/v1alpha1
kind: SecurityProfilesOperatorDaemon
metadata:
  name: spod-a
  namespace: security-profiles-operator
spec:
  enableSelinux: false
  enableLogEnricher: false
  enableAppArmor: false
  nodeSelector:
    node-subset: a
  kubeletRoot: '/foo/bar'
---
apiVersion: security-profiles-operator.x-k8s.io/v1alpha1
kind: SecurityProfilesOperatorDaemon
metadata:
  name: spod-b
  namespace: security-profiles-operator
spec:
  enableSelinux: false
  enableLogEnricher: false
  enableAppArmor: false
  nodeSelector:
    node-subset: b
  kubeletRoot: '/bar/baz'
```

This would create two daemonsets with different `nodeSelector` which would schedule the daemon in different subsets. If we want to take it a step further we could optionally also expose the nodeAffinity to allow users to skip the daemonset scheduler [1].

The advantage of this is that it would finally allow us to cater for mixed OS clusters. e.g. having SPOd instances with SELinux enabled co-existing with instances with AppArmor enabled.

thoughts?

[1] https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#scheduled-by-default-scheduler

**Comment by ccojocar** [comment]
> We could instead have the kubelet folder location be a paramter in the SPOD object, and add the ability to schedule the daemonset in a subset of nodes via a custom selector.

@JAORMX Defining multiple SPOd daemonsets is what I suggested initially in this issue, but in order to map each SPOD CR to a set of nodes, those nodes need to have some sort of selector. The selector can be configured in the SPOD since the affinity is already exposed in the configuration, but this approach has a few challenges in the current setup:
* the operator will have to reconcile multiple SPOD CRs with multiple daemonsets, this will require more code refactoring to make it possible.
* the cluster operators will have to keep in sync manually the operator SPOD CRs definition with the cluster state, for instance every time a new node pool is added to a cluster with a different configuration. If we consider that they need to manage multiple cluster, this can easily get out of control.
* administer the operator with multiple daemonsets it will more complicated, if we consider that we need to monitor that the operator has properly reconciled each each SPOD CR with the corresponding daemonsets.

I think this approach could work in a mid-term, but I opted for a simpler solution for a blocking issue in our case.
